File: attendance/src/main/java/com/shalzz/attendance/ui/timetable/TimeTablePagerFragment.java
/*
 * Copyright (c) 2013-2016 Shaleen Jain <shaleen.jain95@gmail.com>
 *
 * This file is part of UPES Academics.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.shalzz.attendance.ui.timetable;

import android.app.DatePickerDialog;
import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import com.bugsnag.android.Bugsnag;
import com.github.amlcurran.showcaseview.ShowcaseView;
import com.github.amlcurran.showcaseview.targets.Target;
import com.google.android.gms.analytics.HitBuilders;
import com.google.android.gms.analytics.Tracker;
import com.shalzz.attendance.R;
import com.shalzz.attendance.ui.main.MainActivity;
import com.shalzz.attendance.wrapper.DateHelper;

import java.util.Calendar;
import java.util.Date;

import javax.inject.Inject;
import javax.inject.Named;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnPageChange;
import butterknife.Unbinder;

public class TimeTablePagerFragment extends Fragment implements TimeTableMvpView {

    @BindView(R.id.pager)
    public ViewPager mViewPager;

    @Inject @Named("app")
    Tracker mTracker;

    @Inject
    TimeTablePresenter mTimeTablePresenter;

    private int mPreviousPosition = 15;
    private TimeTablePagerAdapter mAdapter;
    private Context mContext;
    private ActionBar actionbar;
    private Unbinder unbinder;

    @Override
    public void onStart() {
        super.onStart();
	    mTracker.setScreenName(getClass().getSimpleName());
        mTracker.send(new HitBuilders.ScreenViewBuilder().build());
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        if (container == null)
            return null;
        mContext = getActivity();
        Bugsnag.setContext("Timetable");
        ((MainActivity) getActivity()).activityComponent().inject(this);

        setHasOptionsMenu(true);
        setRetainInstance(false);
        final View view = inflater.inflate(R.layout.fragment_viewpager, container, false);
        unbinder = ButterKnife.bind(this, view);
        mTimeTablePresenter.attachView(this);

        actionbar = ((AppCompatActivity) getActivity()).getSupportActionBar();

        mAdapter = new TimeTablePagerAdapter(getChildFragmentManager(), mContext);

        mViewPager.setOffscreenPageLimit(3);
        mViewPager.setAdapter(mAdapter);

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        showcaseView();
    }

    public void showcaseView() {
        final ShowcaseView sv = new ShowcaseView.Builder(getActivity())
                .setStyle(R.style.ShowcaseTheme)
                .setTarget(Target.NONE)
                .singleShot(3333)
                .doNotBlockTouches()
                .setContentTitle(getString(R.string.sv_timetable_title))
                .setContentText(getString(R.string.sv_timetable_content))
                .build();

        sv.overrideButtonClick(v -> sv.hide());
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater menuInflater) {
        menuInflater.inflate(R.menu.time_table, menu);
    }

    /* Called whenever we call invalidateOptionsMenu() */
    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        updateTitle(-1);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
         if (item.getItemId() == R.id.menu_date) {
            Calendar today = Calendar.getInstance();
            today.setTime(new Date());
            DatePickerDialog mDatePickerDialog = new DatePickerDialog(mContext, onDateSetListener()
                    , today.get(Calendar.YEAR)
                    , today.get(Calendar.MONTH)
                    , today.get(Calendar.DAY_OF_MONTH));
            mDatePickerDialog.show();

            mTracker.send(new HitBuilders.EventBuilder()
                    .setCategory("Action")
                    .setAction("Scroll to Date")
                    .build());
            return true;
        } else if (item.getItemId() == R.id.menu_today) {
            setDate(new Date());

            mTracker.send(new HitBuilders.EventBuilder()
                    .setCategory("Action")
                    .setAction("Scroll to Today")
                    .build());
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    /**
     * Update action bar title and subtitle
     *
     * @param position to update for, -1 for current page
     */
    @OnPageChange(R.id.pager)
    public void updateTitle(int position) {
        if (position > 0)
            mPreviousPosition = position;
        Date mDate = mAdapter.getDateForPosition(mPreviousPosition);
        if (mDate != null) {
            actionbar.setTitle(DateHelper.getProperWeekday(mDate));
            actionbar.setSubtitle(DateHelper.formatToProperFormat(mDate));
        }
    }

    DatePickerDialog.OnDateSetListener onDateSetListener() {
        return (view, year, monthOfYear, dayOfMonth) -> {
            Calendar date = Calendar.getInstance();
            date.set(year, monthOfYear, dayOfMonth);
            setDate(date.getTime());
            mTracker.send(new HitBuilders.EventBuilder()
                    .setCategory("Scroll to Date")
                    .setAction("Button")
                    .setLabel("OK")
                    .setValue(date.getTimeInMillis())
                    .build());
        };
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
        mTimeTablePresenter.detachView();
    }

    /******* MVP View methods implementation *****/

    @Override
    public void setDate(Date date) {
        mAdapter.setDate(date);
        updateTitle(-1);
        scrollToDate(date);
    }

    @Override
    public void scrollToDate(Date date) {
        int pos = mAdapter.getPositionForDate(date);
        mViewPager.setCurrentItem(pos, true);
    }
}

File: app/src/main/java/com/shalzz/attendance/ui/timetable/TimeTablePagerFragment.kt
/*
 * Copyright (c) 2013-2018 Shaleen Jain <shaleen.jain95@gmail.com>
 *
 * This file is part of College Academics.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.shalzz.attendance.ui.timetable

import android.content.Context
import android.os.Bundle
import android.view.*
import androidx.appcompat.app.ActionBar
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import androidx.viewpager.widget.ViewPager
import com.bugsnag.android.Bugsnag
import com.github.amlcurran.showcaseview.ShowcaseView
import com.github.amlcurran.showcaseview.targets.Target
import com.google.android.material.datepicker.MaterialDatePicker
import com.google.firebase.analytics.FirebaseAnalytics
import com.shalzz.attendance.R
import com.shalzz.attendance.databinding.FragmentViewpagerBinding
import com.shalzz.attendance.ui.main.MainActivity
import com.shalzz.attendance.utils.RxEventBus
import com.shalzz.attendance.wrapper.DateHelper
import java.util.*
import javax.inject.Inject
import javax.inject.Named

class TimeTablePagerFragment : Fragment(), TimeTableMvpView {

    private lateinit var mViewPager: ViewPager

    @Inject
    @field:Named("app")
    lateinit var mTracker: FirebaseAnalytics

    @Inject
    lateinit var mTimeTablePresenter: TimeTablePresenter

    @Inject
    lateinit var mActivity: AppCompatActivity

    @Inject
    lateinit var eventBus: RxEventBus

    private var mPreviousPosition = 15
    private var mAdapter: TimeTablePagerAdapter? = null
    private var mContext: Context? = null
    private var actionbar: ActionBar? = null

    private var _binding: FragmentViewpagerBinding? = null
    // This property is only valid between onCreateView and
    // onDestroyView.
    private val binding get() = _binding!!

    private class OnPageChangeListener(private val callback: (position: Int) -> Unit )
        : ViewPager.SimpleOnPageChangeListener() {

        override fun onPageSelected(position: Int) {
            callback(position)
        }
    }

    override fun onStart() {
        super.onStart()
        mTracker.setCurrentScreen(mActivity, javaClass.simpleName, javaClass.simpleName)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        if (container == null)
            return null
        mContext = activity
        Bugsnag.setContext("Timetable")
        (activity as MainActivity).activityComponent().inject(this)

        setHasOptionsMenu(true)
        retainInstance = false
        _binding = FragmentViewpagerBinding.inflate(inflater, container, false)
        val view = binding.root
        mViewPager = binding.pager

        actionbar = (activity as AppCompatActivity).supportActionBar

        mAdapter = TimeTablePagerAdapter(childFragmentManager,
                mActivity,
                { position -> mViewPager.setCurrentItem(position, true) },
                eventBus)

        mViewPager.offscreenPageLimit = 3
        mViewPager.adapter = mAdapter
        mViewPager.addOnPageChangeListener( OnPageChangeListener(this::updateTitle) )

        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        mAdapter!!.scrollToToday()
    }

    override fun onResume() {
        super.onResume()
        showcaseView()
    }

    private fun showcaseView() {
        val sv = ShowcaseView.Builder(requireActivity())
                .setStyle(R.style.ShowcaseTheme)
                .setTarget(Target.NONE)
                .singleShot(3333)
                .doNotBlockTouches()
                .setContentTitle(getString(R.string.sv_timetable_title))
                .setContentText(getString(R.string.sv_timetable_content))
                .build()

        sv.overrideButtonClick { sv.hide() }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.time_table, menu)
    }

    /* Called whenever we call invalidateOptionsMenu() */
    override fun onPrepareOptionsMenu(menu: Menu) {
        updateTitle(-1)
        super.onPrepareOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.menu_date) {
            val today = Date()
            val datePickerBuilder = MaterialDatePicker.Builder.datePicker()
            datePickerBuilder.setSelection(today.time)
            val datePicker = datePickerBuilder.build()
            datePicker.addOnPositiveButtonClickListener { selection: Long? ->
                setDate(Date(selection!!))
            }
            datePicker.show(getParentFragmentManager(), datePickerBuilder.toString())

            val bundle = Bundle()
            bundle.putString(FirebaseAnalytics.Param.ITEM_ID, item.title.toString())
            bundle.putString(FirebaseAnalytics.Param.ITEM_NAME, "Scroll to Date")
            bundle.putString(FirebaseAnalytics.Param.CONTENT_TYPE, "menu")
            mTracker.logEvent(FirebaseAnalytics.Event.SELECT_ITEM, bundle)
            return true
        } else if (item.itemId == R.id.menu_today) {
            setDate(Date())

            val bundle = Bundle()
            bundle.putString(FirebaseAnalytics.Param.ITEM_ID, item.title.toString())
            bundle.putString(FirebaseAnalytics.Param.ITEM_NAME, "Scroll to Today")
            bundle.putString(FirebaseAnalytics.Param.CONTENT_TYPE, "menu")
            mTracker.logEvent(FirebaseAnalytics.Event.SELECT_ITEM, bundle)
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    /**
     * Update action bar title and subtitle
     *
     * @param position to update for, -1 for current page
     */
    private fun updateTitle(position: Int) {
        if (position != -1)
            mPreviousPosition = position
        val mDate = mAdapter!!.getDateForPosition(mPreviousPosition)
        if (mDate != null) {
            actionbar!!.title = DateHelper.getProperWeekday(mDate)
            actionbar!!.subtitle = DateHelper.toProperFormat(mDate)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        mTimeTablePresenter.detachView()
        mAdapter!!.destroy()
    }

    /******* MVP View methods implementation  */

    override fun setDate(date: Date) {
        mAdapter!!.setDate(date)
        mAdapter!!.scrollToDate(date)
        updateTitle(-1)
    }

}
